### PR TITLE
USB Middleware

### PR DESCRIPTION
**Pull request by liam-ilan**
### Changelist 
- Added USB middleware flag to stm32 h7 and f4 binary building functions.
- Setup USB Echo Program on F4 and H7 (H7 not tested due to clock misconfiguration).

### Testing Done
- USB Device Appears as device when plugged into computer (f4dev).
- Able to communicate with f4dev via pyusb (@setaremalekiii).
- h7dev untested, to be done when clock configs fixed.
- Ran subsequent `hw_usb` library on top of this one, usb comms working awesome.

### Resolved Tickets
[FIRM-159](https://ubcformulaelectric.atlassian.net/browse/FIRM-159)

[FIRM-159]: https://ubcformulaelectric.atlassian.net/browse/FIRM-159?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: firmware/shared/src/app/app_utils.h [hunk 1]
```diff
@@ -6,8 +6,14 @@
...
 
-#define MIN(x, y) (((x) < (y)) ? (x) : (y))
-#define MAX(x, y) (((x) > (y)) ? (x) : (y))
+#ifndef MIN
+#define MIN(a, b) (((a) < (b)) ? (a) : (b))
+#endif
+
+#ifndef MAX
+#define MAX(a, b) (((a) > (b)) ? (a) : (b))
+#endif
```
Comment (Lucien950) [inline firmware/shared/src/app/app_utils.h:15]: did something happen?

**Review by Lucien950 — CHANGES_REQUESTED** [review]
I am lowkey leaning on not using the STM generated stuff for CDC level of USB at least cause it seems very big and confusing.

Is the delieverable here to have any communication between USB and STM? I assume in a later PR you will do the migration for Chimera from UART to USB.

Overall very good work, and I look forward to the demo on software-dubs

File: firmware/dev/h7dev/CMakeLists.txt [hunk 2]
```diff
@@ -101,6 +105,7 @@ IF ("${TARGET}" STREQUAL "binary")
     target_link_libraries("h7dev_app.elf" "h7dev_stm32cube" "protobuf_library" "logfs_cm7")
 
     # Build bootloader.
+    message("boottttt")
```
[the remark this answers is not in the document]
Comment (liam-ilan) [reply firmware/dev/h7dev/CMakeLists.txt:108]: Fixed 😅

File: firmware/shared/src/app/app_utils.h (hunk 1 above)
Comment (liam-ilan) [reply firmware/shared/src/app/app_utils.h:15]: STM defined it's own min and max macros :/. Painful necessity.

File: firmware/dev/f4dev/cubemx/Src/usbd_cdc_if.c [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+static int8_t CDC_Receive_FS(uint8_t *Buf, uint32_t *Len)
+{
+    /* USER CODE BEGIN 6 */
+    USBD_CDC_SetRxBuffer(&hUsbDeviceFS, &Buf[0]);
+    USBD_CDC_ReceivePacket(&hUsbDeviceFS);
+
+    // Test: Send data back
+    uint16_t len = (uint16_t)*Len;
+    CDC_Transmit_FS(Buf, len);
+
```
[the remark this answers is not in the document]
Comment (liam-ilan) [reply firmware/dev/f4dev/cubemx/Src/usbd_cdc_if.c:270]: This PR is just setting up the build system and middleware, @setaremalekiii is working on the wrapper. Labeled as a test.

File: firmware/dev/f4dev/CMakeLists.txt [hunk 4]
```diff
@@ -36,6 +36,10 @@ IF ("${TARGET}" STREQUAL "binary")
             "stm32f4xx_hal.c"
             "stm32f4xx_hal_sd.c"
             "stm32f4xx_ll_sdmmc.c"
+            "stm32f4xx_hal.c"
```
[the remark this answers is not in the document]
Comment (liam-ilan) [reply firmware/dev/f4dev/CMakeLists.txt:39]: "stm32f4xx_hal.c" was a dupe, but the other hal libraries are needed for USB - My gut punch tells me to add them to the `stm32f412rx_cube_library` and stm32h712rx_cube_library` functions, but in that case it dosen't make sense to fragment the HAL includes between the individual board and shared cmake - lmk if you want me to move all of these into `stm32____rx_cube_library`

File: firmware/cmake/embedded.cmake [hunk 5]
```diff
@@ -40,7 +40,6 @@ set(SHARED_COMPILER_FLAGS
         --specs=nosys.specs
         --specs=nano.specs
         -Wall
-        -Werror
```
[the remark this answers is not in the document]
Comment (liam-ilan) [reply firmware/cmake/embedded.cmake:43]: Fixed!

File: firmware/cmake/stmlib.cmake [hunk 6]
```diff
@@ -80,9 +83,20 @@ function(stm32f412rx_cube_library
...
+    endif()
+
+    file(GLOB USB_SRCS
+        "${USB_MIDDLEWARE_DIR}/Class/CDC/Src/usbd_cdc.c"
+        
+        "${USB_MIDDLEWARE_DIR}/Core/Src/usbd_core.c"
+        "${USB_MIDDLEWARE_DIR}/Core/Src/usbd_ctlreq.c"
+        "${USB_MIDDLEWARE_DIR}/Core/Src/usbd_ioreq.c"
+    )
+
```
[the remark this answers is not in the document]
Comment (liam-ilan) [reply firmware/cmake/stmlib.cmake]: Done!

File: firmware/cmake/stmlib.cmake [hunk 7]
```diff
@@ -109,6 +123,10 @@ function(stm32f412rx_cube_library
     set(STARTUP_SRC "${DRIVERS_DIR}/CMSIS/Device/ST/STM32F4xx/Source/Templates/gcc/startup_stm32f412rx.s")
 
     set(STM32CUBE_SRCS ${STM32_HAL_SRCS} ${RTOS_SRCS} ${SYSTEMVIEW_SRCS} ${SYSCALLS} ${IOC_CHECKSUM} ${STARTUP_SRC} ${NEWLIB_SRCS})
+    if(USB_ENABLED)
+        list(APPEND STM32CUBE_SRCS ${USB_SRCS})
+    endif()
+
```
[the remark this answers is not in the document]
Comment (liam-ilan) [reply firmware/cmake/stmlib.cmake]: Done!

File: firmware/cmake/stmlib.cmake [hunk 8]
```diff
@@ -130,25 +148,33 @@ function(stm32h733xx_cube_library
...
+        # SEGGER SystemView includes.
+        "${SEGGER_SYSTEMVIEW_SOURCE_DIR}/SEGGER"
+        "${SEGGER_SYSTEMVIEW_SOURCE_DIR}/Config"
+        "${SEGGER_SYSTEMVIEW_SOURCE_DIR}/Sample/FreeRTOSV10"
     )
+
+    if(USB_ENABLED)
+        list(APPEND STM32CUBE_INCLUDE_DIRS "${USB_MIDDLEWARE_DIR}/Core/Inc" "${USB_MIDDLEWARE_DIR}/Class/CDC/Inc")
+    endif()
```
[the remark this answers is not in the document]
Comment (liam-ilan) [reply firmware/cmake/stmlib.cmake]: Done!

File: firmware/cmake/stmlib.cmake [hunk 9]
```diff
@@ -81,9 +84,16 @@ function(stm32f412rx_cube_library
             "${SEGGER_SYSTEMVIEW_SOURCE_DIR}/Sample/FreeRTOSV10"
     )
 
+    file(GLOB USB_SRCS
```
Comment (Lucien950) [inline firmware/cmake/stmlib.cmake:87]: no glob required

please move to line 126

File: firmware/cmake/stmlib.cmake [hunk 10]
```diff
@@ -54,15 +54,18 @@ message("  🔃 Registered stm32f412rx_cube_library() function")
...
         HAL_LIB_NAME
         HAL_CONF_DIR
         HAL_SRCS
         SYSCALLS
         IOC_CHECKSUM
+        USB_ENABLED
 )
     set(DRIVERS_DIR "${STM32CUBEF4_SOURCE_DIR}/Drivers")
     set(FREERTOS_DIR "${STM32CUBEF4_SOURCE_DIR}/Middlewares/Third_Party/FreeRTOS/Source")
+    set(USB_MIDDLEWARE_DIR "${STM32CUBEF4_SOURCE_DIR}/Middlewares/ST/STM32_USB_Device_Library")
```
Comment (Lucien950) [inline firmware/cmake/stmlib.cmake:68]: move to line 126

File: firmware/cmake/stmlib.cmake [hunk 11]
```diff
@@ -133,26 +157,29 @@ function(stm32h733xx_cube_library
...
-            "${HAL_CONF_DIR}"
-            "${DRIVERS_DIR}/STM32H7xx_HAL_Driver/Inc"
-            "${DRIVERS_DIR}/STM32H7xx_HAL_Driver/Inc/Legacy"
-            "${FREERTOS_DIR}/include"
-            "${FREERTOS_DIR}/CMSIS_RTOS_V2"
-            "${FREERTOS_DIR}/portable/GCC/ARM_CM4F"
-            "${DRIVERS_DIR}/CMSIS/Device/ST/STM32H7xx/Include"
-            "${DRIVERS_DIR}/CMSIS/Include"
-            # SEGGER SystemView includes.
-            "${THIRD_PARTY_DIR}/sysview"
```
Comment (Lucien950) [inline firmware/cmake/stmlib.cmake:151]: word

File: firmware/cmake/stmlib.cmake [hunk 12]
```diff
@@ -133,26 +157,29 @@ function(stm32h733xx_cube_library
         HAL_SRCS
         SYSCALLS
         IOC_CHECKSUM
+        USB_ENABLED
 )
     set(DRIVERS_DIR "${STM32CUBEH7_SOURCE_DIR}/Drivers")
     set(FREERTOS_DIR "${STM32CUBEH7_SOURCE_DIR}/Middlewares/Third_Party/FreeRTOS/Source")
+    set(USB_MIDDLEWARE_DIR "${STM32CUBEH7_SOURCE_DIR}/Middlewares/ST/STM32_USB_Device_Library")
```
Comment (Lucien950) [inline firmware/cmake/stmlib.cmake:164]: similar points as above

File: firmware/cmake/stmlib.cmake (hunk 9 above)
Comment (Lucien950) [reply firmware/cmake/stmlib.cmake:87]: actually why this cause down there it's the same files

File: firmware/cmake/stmlib.cmake [hunk 13]
```diff
@@ -54,15 +54,18 @@ message("  🔃 Registered stm32f412rx_cube_library() function")
...
 # HAL_SRCS: the ones that we want, stripped prefixes
 # SYSCALLS: most of the functions defined inside are weak references, only used to make sure it builds without error.
+# USB_ENABLED: flags if usb middleware should be included.
 function(stm32f412rx_cube_library
         HAL_LIB_NAME
         HAL_CONF_DIR
         HAL_SRCS
         SYSCALLS
         IOC_CHECKSUM
+        USB_ENABLED
```
Comment (Lucien950) [inline firmware/cmake/stmlib.cmake:64]: they all gotta be enabled no? cause chimera everywhere

**Review by Lucien950 — CHANGES_REQUESTED** [review]
last few things

File: firmware/cmake/stmlib.cmake (hunk 13 above)
Comment (liam-ilan) [reply firmware/cmake/stmlib.cmake:64]: I want to support boards that don't have USB (currently thats quadruna, but future dev boards maybe? I guess on quintuna this is true but down the line it probably isnt...)

Comment (Lucien950) [reply firmware/cmake/stmlib.cmake:64]: just give them usb it's like very short on the compile times?

**Review by Lucien950 — APPROVED** [review]

**Pull request merged**